src/Xamarin.Android.Build.Tasks/Tasks/BuildApk.cs

```diff
@@ -459,10 +462,21 @@ HashSet<string> ParseProfilers (string value)
 			return results;
 		}
 
-		void AddNativeLibrary (ZipArchiveEx apk, string abi, string filename)
+		void AddNativeLibrary (ZipArchiveEx apk, string abi, string filename, string inArchiveFileName = null)
 		{
-			var path    = Path.Combine (MSBuildXamarinAndroidDirectory, "lib", abi, filename);
-			apk.Archive.AddEntry (string.Format ("lib/{0}/{1}", abi, filename), File.OpenRead (path));
+			string libPath = Path.Combine (MSBuildXamarinAndroidDirectory, "lib", abi);
+			string path    = null;
+			if (PreferNativeLibrariesWithDebugSymbols) {
+				string debugPath = Path.Combine (libPath, Path.ChangeExtension (filename, ".d.so"));
+				if (File.Exists (debugPath))
+					path = debugPath;
+			}
+			if (path == null)
```

Comment:
I don't see why you can't just assign `path` at the declaration, and re-assign within the `PreferNativeLibrariesWithDebugSymbols`. It would remove the need for this null check.